storage/engine/gc.go

```diff
@@ -28,84 +27,67 @@ import (
 // policy allows either the union or intersection of maximum # of
 // versions and maximum age.
 type GarbageCollector struct {
-	now      proto.Timestamp // time at start of GC
-	policyFn func(key proto.Key) *proto.GCPolicy
+	expiration proto.Timestamp
+	policy     proto.GCPolicy
 }
 
-// NewGarbageCollector allocates and returns a new GC.
-func NewGarbageCollector(now proto.Timestamp, policyFn func(key proto.Key) *proto.GCPolicy) *GarbageCollector {
+// NewGarbageCollector allocates and returns a new GC, with expiration
+// computed based on current time and policy.TTLSeconds.
+func NewGarbageCollector(now proto.Timestamp, policy proto.GCPolicy) *GarbageCollector {
+	ttlNanos := int64(policy.TTLSeconds) * 1E9
 	return &GarbageCollector{
-		now:      now,
-		policyFn: policyFn,
+		expiration: proto.Timestamp{WallTime: now.WallTime - ttlNanos},
+		policy:     policy,
 	}
 }
 
-// MVCCPrefix returns the full key as prefix for non-version MVCC
-// keys and otherwise just the encoded key portion of version MVCC keys.
-func (gc *GarbageCollector) MVCCPrefix(key proto.EncodedKey) int {
-	remaining, _ := encoding.DecodeBinary(key)
-	return len(key) - len(remaining)
-}
-
 // Filter makes decisions about garbage collection based on the
 // garbage collection policy for batches of values for the same key.
-// The GC policy is determined via the policyFn specified when the
-// GarbageCollector was created. Returns a slice of deletions, one
-// per incoming keys. If an index in the returned array is set to
-// true, then that value will be garbage collected.
-func (gc *GarbageCollector) Filter(keys []proto.EncodedKey, values [][]byte) []bool {
-	if len(keys) == 1 {
-		return nil
-	}
-	toDelete := make([]bool, len(keys))
-	// Decode the first key and make sure it's an MVCC metadata key.
-	dKey, ts, isValue := MVCCDecodeKey(keys[0])
-	if isValue {
-		log.Errorf("unexpected MVCC value encountered: %q", keys[0])
-		return toDelete
+// Returns the timestamp including, and after which, all values should
+// be garbage collected. If no values should be GC'd, returns
+// proto.ZeroTimestamp.
+func (gc *GarbageCollector) Filter(keys []proto.EncodedKey, values [][]byte) proto.Timestamp {
+	if gc.policy.TTLSeconds <= 0 {
+		return proto.ZeroTimestamp
 	}
-	// Using first key, look up the policy which applies to this set of MVCC values.
-	policy := gc.policyFn(dKey)
-	if policy == nil || policy.TTLSeconds <= 0 {
-		return nil
+	if len(keys) == 0 {
+		return proto.ZeroTimestamp
 	}
-	expiration := gc.now
-	expiration.WallTime -= int64(policy.TTLSeconds) * 1E9
 
-	var survivors bool
 	// Loop over remaining values. All should be MVCC versions.
```

Comment:
Delete "remaining" from this comment since this function is no longer responsible for splitting the metadata key from the rest.